Support GitHub Enterprise avatars via Enterprise REST service and optionally Gravatar

kabel opened the pull request:
In GitHub Enterprise, the avatarUrl endpoints that the APIs return require authentication in standard configuration. There is currently no way to do token authentication to these endpoints, so a fallback is necessary to render an image.

The interface was previously coded to handle fallback to the Gravatar service (just like GitHub Desktop [did](https://github.com/desktop/desktop/pull/3513)), but this behavior was disabled in 13ab0a7653008e8433cbac0136d4d8fb97e1328c and 6769dd4cba8b3b6bfef20c94ef28918103634680 because of end-user preference for making this feature configurable.

Recently an enterprise-only REST API was discovered that allows us to proxy the avatar image. This allows us to fetch the binary and encode it as a data URI to inject into the various views that otherwise render the avatar with standard http requests. If this API fails for some reason, Gravatar remains an optional fallback choice.

A new setting `githubPullRequests.defaultGravatarsStyle` has been added of type `enum` with values for each of the Gravatar supported fallback styles. Setting this value to `none` (the default) disables Gravatar integration.

These settings provide a richer user experience for GitHub Enterprise pull requests, allowing users to opt-in/out on avatar rendering via a third-party service with multiple options for the default fallback image style.

For consistency in multiple interface areas where `<Avatar>` is rendered, the GraphQL queries (and matching typescript interfaces) have been updated to include the necessary `email` property for integration with Gravatar.

REST provided actors are missing email. If missing email and using GitHub Enterprise and it's needed by Gravatar, make another REST API call to the user endpoint to get email. This allows a fallback avatar to be generated.

This fixes GitHub Enterprise avatars not being rendered in the tree views too.

The utility functions have been adjusted to support doing the fallback image URL generation directly after the raw results have been loaded from the normal REST and GraphQL APIs. This should be a cleaner implementation.

kabel commented:
This is a long overdue PR [requested](https://github.com/microsoft/vscode-pull-request-github/issues/2771#issuecomment-877267773) by @alexr00

kabel commented:
Example Overview render with default settings:
![image](https://user-images.githubusercontent.com/675956/216652060-1a4ea75e-304d-4db2-923b-be5885acd552.png)

kabel commented:
Looks like I'm referencing outdated information. GitHub Desktop now appears to be using an undocumented GitHub Enterprise REST API endpoint at `/api/v3/enterprise/avatars/u/e?email=`

![image](https://user-images.githubusercontent.com/675956/216659906-ec85bb63-feb5-407e-acb8-5731075e913b.png)

alexr00 commented on src/github/graphql.ts:
```diff
@@ -117,15 +120,26 @@ export interface Commit {
+			avatarUrl: string
+			name: string
+			email: string
+			user: {
+				login: string;
+				avatarUrl: string;
+				url: string;
+				email: string;
```
Same as above: we're now getting all this info twice: once on the `commiter` property and once on the `commiter.user` property.

kabel replied:
See response above. However, it's worth calling out that the `committer` field appears to be unused entirely in the code base/UI.

alexr00 commented on src/github/graphql.ts:
```diff
@@ -680,7 +703,7 @@ export interface GetChecksResponse {
 					commit: {
 						statusCheckRollup?: {
-							state: string;
+							state: 'EXPECTED' | 'ERROR' | 'FAILURE' | 'PENDING' | 'SUCCESS';
```
Once again, I do think this is probably a good change, but to keep what needs to be tested for it scoped down could you revert this?

kabel replied:
Yes. Again, just trying to match it up to the current GraphQL schema.

alexr00 commented on src/github/graphql.ts:
```diff
@@ -612,10 +631,11 @@ export interface ContributionsCollection {
-		avatarUrl?: string;
-		bio?: string;
-		company?: string;
-		location?: string;
+		avatarUrl: string;
+		email: string;
+		bio: string;
+		company: string;
```
The `?` on these is correct. A user doesn't have to have these set.

```suggestion
		avatarUrl?: string;
		email?: string;
		bio?: string;
		company?: string;
```

kabel replied:
A user doesn't have to set them, but that will make them null/empty, not undefined/missing, in the API response. From the GraphQL schema, these fields will always return some value. It's one of those cases where null vs undefined mean very different things. Unless I'm reading the schema wrong or there have been schema changes?

![image](https://user-images.githubusercontent.com/675956/217588629-0d71d65d-1804-4818-9366-d107839e8fd7.png)

![image](https://user-images.githubusercontent.com/675956/217588590-5e20248c-3fb3-4d7d-87fe-cbbae30a7337.png)

![image](https://user-images.githubusercontent.com/675956/217588528-0865c494-21a0-43dc-b289-e5873bda9a1f.png)

![image](https://user-images.githubusercontent.com/675956/217588428-d125f927-4e15-40f0-b4f8-2ff3e578923e.png)

Example response:
![image](https://user-images.githubusercontent.com/675956/217589376-1b81628e-a97f-4021-aab0-64d8700efc80.png)

kabel replied:
I'll change them to `string | null` as the schema defines them.

kabel commented on package.json:
```diff
@@ -503,6 +503,36 @@
+					"description": "%githubPullRequests.githubEnterpriseAvatars.description%",
+					"scope": "window"
+				},
+				"githubPullRequests.defaultGravatarsStyle": {
```
Good question. I took inspiration from Git Lens that has these settings separated because of multiple implementations. I'd be fine either way.

kabel commented on src/github/graphql.ts:
```diff
@@ -117,15 +120,26 @@ export interface Commit {
 	id: string;
 	commit: {
 		author: {
+			avatarUrl: string
+			name: string
+			email: string
```
It's possible for the author and/or committer to not be _GitHub_ users, but rather just standard git users. In those cases the email will come from the commit, rather than the GitHub profile.

kabel commented on src/github/utils.ts:
```diff
@@ -766,7 +775,7 @@ export function parseGraphQLTimelineEvents(
 					author: commitEv.commit.author.user
 						? parseAuthor(commitEv.commit.author.user, githubRepository)
-						: { login: commitEv.commit.committer.name },
+						: { login: commitEv.commit.author.name, avatarUrl: commitEv.commit.author.avatarUrl, email: commitEv.commit.author.email },
```
Good question. It's difficult to understand the UI purpose of this model, but it doesn't really make sense to change the meaning/display of `author` to `committer` only when the `author` is not a _GitHub_ user. Non-GitHub authors _should_ be treated/displayed just like any other user.

kabel commented:
Thanks for the review/questions. 

Based on the availability of the Enterprise REST API (as now used in GitHub Desktop), I added a bit of scope creep to this request. In c0145681df76ef688929ab21181245974db7725e the fallback logic will first check the REST API for the avatar image binary. If the API returns the image binary, it is converted to a data URI to be rendered by the webviews and treeviews. I updated the CSP meta of the webviews to allow `img-src data:` to make that work. If that's too much for this PR, I can save that commit for a future request.

Other comments that don't require additional feedback should now be addressed.

kabel commented on src/github/utils.ts:
```diff
@@ -1110,13 +1109,98 @@ export function hasEnterpriseUri(): boolean {
-	return authProviderId === AuthProvider.github ? avatarUrl : (email ? generateGravatarUrl(
-		crypto.createHash('md5').update(email?.trim()?.toLowerCase()).digest('hex')) : undefined);
+const enterpriseAvatarQueue = new PQueue({concurrency: 3});
+const enterpriseAvatarCache: {[k: string]: Promise<string | void>} = {};
```
This only caches enterprise avatars in memory. This may not be desirable if we want to keep memory usage low. There is an opportunity to improve this static-asset caching by saving these images into the extension's storage URI on the filesystem and then having the webviews serve them from that `vscode-resource:` location. That would avoid having to re-fetch the avatars every time the window reloads.

alexr00 commented on src/github/utils.ts:
```diff
@@ -1110,13 +1109,98 @@ export function hasEnterpriseUri(): boolean {
+
+		if (!(cacheKey in enterpriseAvatarCache)) {
+			enterpriseAvatarCache[cacheKey] = enterpriseAvatarQueue.add(() =>
+				octokit.api.request(`GET ${enterpriseAvatarRestBase}${avatarUri.path}`, options).then(
+					resp => {
+						const dataUri = `data:${resp.headers['content-type']};base64,${Buffer.from(resp.data).toString(
+							'base64',
+						)}`;
+						return dataUri;
+					},
+					() => {
+						delete enterpriseAvatarCache[cacheKey];
+						return;
+					},
+				),
```
Why is this needed?

kabel replied:
See https://github.com/microsoft/vscode-pull-request-github/pull/4497#issuecomment-1423080477

alexr00 commented:
I'm sorry @kabel but this PR has become too big for me to continue reviewing now and it mostly seems to be because of fetching the user email when parsing user results from the GitHub REST API. 

I will try to plan time to review in the next couple months, but ideally the PR would be much smaller.

Thank you again for this change. If I hear from users that gravatars are important I will definitely reference your changes to see how to enable them.

alexr00 commented:
I've opened an issue to try to gauge how much desire there is for this feature: https://github.com/microsoft/vscode-pull-request-github/issues/4517

kabel commented on src/github/activityBarViewProvider.ts:
```diff
@@ -382,7 +386,7 @@ export class PullRequestViewProvider extends WebviewViewBase implements vscode.W
 	<head>
 		<meta charset="UTF-8">
-		<meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src vscode-resource: https:; script-src 'nonce-${nonce}'; style-src vscode-resource: 'unsafe-inline' http: https: data:;">
+		<meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src vscode-resource: https: data:; script-src 'nonce-${nonce}'; style-src vscode-resource: 'unsafe-inline' http: https: data:;">
```
It's part of the scope creep I mentioned at https://github.com/microsoft/vscode-pull-request-github/pull/4497#issuecomment-1423080477.

I modified this PR to include grabbing the avatar binary from

1. the (undocumented) GitHub Enterprise REST API for avatars
2. the Gravatar service

The former is really the preferred method of getting avatars as it uses the true system-of-record for the user's preferred image. The latter is an opinion when all else fails. Currently, I have the former integrated by encoding the avatar API response (binary image) into a data URI.

![image](https://user-images.githubusercontent.com/675956/218143104-edfff3e5-30a4-4405-a55e-34c26a13f38f.png)

kabel commented on src/github/utils.ts:
```diff
@@ -219,14 +221,26 @@ export function updateCommentReviewState(thread: GHPRCommentThread, newDraftMode
 }
 
-export function convertRESTUserToAccount(
+export async function convertRESTUserToAccount(
```
Following your advice, I refactored and force-pushed the implementation to be much cleaner and simpler.

kabel commented on src/github/githubRepository.ts:
```diff
@@ -509,21 +514,28 @@ export class GitHubRepository implements vscode.Disposable {
+							return { milestone, issues };
+						}),
+					)
+				).filter(x => x !== null) as { milestone: IMilestone; issues: IssueModel[] }[];
```
This was based on how other filtering loops were done in the file, but has since been refactored.

kabel commented:
Sounds fair @alexr00. I've got the latest push down to a more manageable size (mostly added lines for the new Enterprise REST integration), but I'll wait for more time to become available. Having a better Enterprise experience continues to be important for me, but I can understand how avatars aren't going to be a huge win feature-wise for folks.
